dot/parachain/dispute/participation.go

```diff
@@ -0,0 +1,372 @@
+package dispute
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	parachain "github.com/ChainSafe/gossamer/dot/parachain/runtime"
+
+	"github.com/ChainSafe/gossamer/dot/parachain/dispute/overseer"
+	"github.com/ChainSafe/gossamer/dot/parachain/dispute/types"
+	parachainTypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+// CandidateComparator comparator for ordering of disputes for candidate.
+type CandidateComparator struct {
+	relayParentBlockNumber *uint32
+	candidateHash          common.Hash
+}
+
+// NewCandidateComparator creates a new CandidateComparator.
+func NewCandidateComparator(relayParentBlockNumber *uint32,
```

Comment:
Could you elaborate on why you'd prefer it that way?

Reply:
I will rephrase myself hahah. So `New...` functions could return an error if an error was found while creating the instance/object, for example `os.Create` return an error if it could not create a file/folder in the OS. However in this case `NewCandidateComparator` receives the `CandidateReceipt` and executes the `Hash()` method which if fails return, but IMO this error is not intrinsic to the instance

Same for `getBlockNumber` function that receives the `CandidateReceipt` and do the same check but could receive only the hash as an argument 

```go
receiptHash, err :=  request.candidateReceipt.Hash()
if err != nil { ... }

blockNumber, err := getBlockNumber(ctx.Sender, receiptHash)
if err != nil {
	return fmt.Errorf("get block number: %w", err)
}

comparator :=  NewCandidateComparator(&blockNumber, receiptHash)
```

Reply:
I agree, thank you!